### DISCUSSION TITLE
Question on using Mojo::Base with the model class on the same file

### DISCUSSION THREAD

**Discussion by vmmello** [mojolicious/mojo#1923] [General]
Hi,

I'm trying learn Mojo::Base and create a simple model class with it starting with a simple test file:

```perl
#!/usr/bin/perl
package main;

use strict;
use warnings;

my $obj = MyModel->new;

$obj->name('whatever');

1;

package MyModel;

use Mojo::Base -base;

has 'name';

1;

```

Though when I run it I get the following error:
```
Can't locate object method "name" via package "MyModel" at ./test.pl line 9.
```

Line 9 is:
```perl
$obj->name('whatever');
```

Though if I move the class `MyModel` to a separate file, it works as expected. It also works as expected if I use `$obj->attr('name' => ...)` instead of `$obj->name(...)`.

So I ask, am I doing anything wrong? Or could the function `has()` in `Mojo::Base` be slightly modified to work on classes defined in the same file?

Tested with `Mojolicious 9.22`

Thanks in advance.

**Comment by Grinnz** [comment]
When defining packages in the same file, ordering becomes an issue, since you don't have the ordering provided by a `use` statement which will load the whole module at parse time. To avoid any possible ordering issue, encase your inline module in a BEGIN block and put that before your code, as you would a `use` statement.

**Reply by vmmello** [reply]
Yes, that did it! Thanks for the detailed explanation!
